### PR TITLE
Fix InvalidCastException in ErrorProvider

### LINKED ISSUE #8424: [.NET 7.0.1] InvalidCastException in ErrorProvider
### .NET version

7.0.1

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

_No response_

### Issue description

System.InvalidCastException exception is thrown in ErrorProvider.

### Steps to reproduce

1. Create a .NET 7.0 WinForm app.
2. Replace Form1.cs with the following code
3. Launch the app.

```
namespace WinFormsApp1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            var model = new MyModel();
            this.DataBindings.Add("Text", model, "Error");
            new ErrorProvider(this).DataSource = model; // InvalidCastException !!!
        }
    }

    public class MyModel : System.ComponentModel.IDataErrorInfo
    {
        public string this[string columnName] => "";
        public string Error => "";
    }
}
```
_Excepted Result:_
No exception thrown.

_Actual Result:_
This exception is thrown in code:

> System.InvalidCastException
>   HResult=0x80004002
>   Message=Unable to cast object of type 'System.Collections.Generic.KeyValuePair`2[System.Windows.Forms.Control,System.String]' to type 'System.Collections.DictionaryEntry'.
>   Source=System.Windows.Forms
>   Stack Trace:
>    at System.Windows.Forms.ErrorProvider.ErrorManager_CurrentChanged(Object sender, EventArgs e)
>    at System.Windows.Forms.ErrorProvider.SetErrorManager(Object newDataSource, String newDataMember, Boolean force)
>    at WinFormsApp1.Form1..ctor() (Form1.cs):line 13
>    at WinFormsApp1.Program.Main() (Program.cs):line 14

**Other information**
https://github.com/dotnet/winforms/pull/7037
```
// Hashtable controlError = new Hashtable(bindingsCount);
Dictionary<Control, string> controlError = new(bindingsCount);
...
IEnumerator enumerator = controlError.GetEnumerator();
while (enumerator.MoveNext())
{
    DictionaryEntry entry = (DictionaryEntry)enumerator.Current; // Invalid Cast !!!
    SetError((Control)entry.Key, (string?)entry.Value);
}
```

### PR DESCRIPTION
**Pull request by gpetrou**
Fixes #8424

## Proposed changes

- Add simple test.
- Fix InvalidCastException in ErrorProvider

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8425)

### CODE REVIEW

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ErrorProviderTests.cs [hunk 1]
```diff
@@ -1244,6 +1244,26 @@ public void ErrorProvider_HasErrors_ControlErrorReset()
             Assert.False(provider.HasErrors);
         }
 
+        [WinFormsFact]
+        public void ErrorProvider_CustomDataSource_DoesNotThrowInvalidCastException()
+        {
+            using Form form = new();
+            var customDataSource = new CustomDataSource();
```
Comment (lonitra) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ErrorProviderTests.cs:1251]: nit:
```suggestion
            CustomDataSource customDataSource = new();
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ErrorProviderTests.cs [hunk 2]
```diff
@@ -1244,6 +1244,26 @@ public void ErrorProvider_HasErrors_ControlErrorReset()
...
+            form.DataBindings.Add("Text", customDataSource, "Error");
+            using var errorProvider = new ErrorProvider(form);
```
Comment (lonitra) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ErrorProviderTests.cs:1253]:
```suggestion
            using ErrorProvider errorProvider = new(form);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ErrorProvider.cs [hunk 3]
```diff
@@ -513,11 +512,11 @@ private void ErrorManager_CurrentChanged(object? sender, EventArgs e)
...
             }
 
-            IEnumerator enumerator = controlError.GetEnumerator();
+            var enumerator = controlError.GetEnumerator();
             while (enumerator.MoveNext())
             {
-                DictionaryEntry entry = (DictionaryEntry)enumerator.Current;
-                SetError((Control)entry.Key, (string?)entry.Value);
+                var entry = enumerator.Current;
+                SetError(entry.Key, entry.Value);
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/ErrorProvider.cs:519]: Could this be changed to the following so the type can be recognized more quickly?
```c#
foreach (KeyValuePair<Control, string> entry in controlError) 
{
   SetError(entry.Key, entry.Value);
}
```

**Review by lonitra — COMMENTED** [review]
Few small comments but otherwise looks good

**Review by lonitra — APPROVED** [review]
Thank you!

**Comment by gpetrou** [comment]
@lonitra perhaps you would like to backport this since it is a regression in .NET 7.

**Comment by lonitra** [comment]
> @lonitra perhaps you would like to backport this since it is a regression in .NET 7.

Yep thanks for calling that out! (: I'll get another pair of eyes to look at the changes here and start backporting process then.

**Review by dreddy-work — APPROVED** [review]

**Comment by dreddy-work** [comment]
@gpetrou , auto port fails with conflicts. Can you create a new PR pointing to release/7.0 branch? I will help driving it from there.

**Comment by dreddy-work** [comment]
manually porting.

**Pull request merged**